### ISSUE TITLE
ConsultREST API not connecting to REST API

### ISSUE THREAD

**Issue by Reis-A** [devture/matrix-corporal#26]
I installed matrix-corporal in a docker container and it is working alright.
Homeserver is installed on docker host and accessible via local ip and port.

I wanted to use the REST API Hook and installed a minimum REST API  with flask on the docker host. The API is running and it reacts to curl get requests curl -Xget http://local-ip:5000/restapi

 from the docker host as well as inside the matrix-corporal container when i use curl.

However, the ConsultRestAPI that i setup in the policy file never connects to the API on local-ip:5000. There is not even a request that arrives at the flask api. 
What am I doing wrong? or why does the http module from go not connect..?

**Comment by spantaleev** [comment]
You need to run both containers on the same
container network and point corporal
to the name and port of the other container, not to localhost

**Comment by Reis-A** [comment]
ok. so installed matrix-corporal 2.3.2 before October 2022 when the file rest-service.go got updated.
 and there the hook system works. so it is definitely a bug somewhere from transitioning from 2.3.2 to 2.40 where some changes where made in the handling of the context in go.
Since I have no clue about go, I cannot be of any help here :)

**Comment by Reis-A** [comment]
My problem with the ConsultREST API still persists.
I did some more tests in the development system I know now, that all works if the code in  the file /corporal/hooks/rest_service_consulting.go in line 262  from 
`ctx, cancel := context.WithTimeout(context.Background(), timeoutDuration)
		defer cancel()`
back to (from matrix-corporal 2.3.2)
`ctx, _ := context.WithTimeout(context.Background(), timeoutDuration)`

I don't really know the difference in the code and what the problem is with the new code... or maybe it is my setup?
but as i said i am using the development environment here...

**Comment by Cybernet1k** [comment]
The same problem arises. Rest service used self-descriptive and hook-rest-service in containers with the same network. The matrix-corporal image is used with https://hub.docker.com/r/devture/matrix-corporal.

**Comment by Rhywden** [comment]
Yeah, for me both are in the same network. The rest-service is reachable under "http://api:8080/" (I tested this from within the containers using curl) but whatever I do, the hook just yields this error:

```
time="2025-03-06T10:32:11Z" level=warning msg="Async REST service suffered an error: Failed after trying 1 times. Last error: Error fetching from URL: Post \"http://api:8080/corporal/logeverything\": context canceled" hookChain="#log_everything" hookEventType=beforeAnyRequest hookId=log_everything method=POST uri=/_matrix/client/r0/keys/query
```

**Comment by Reis-A** [comment]
i looked at this cancellation of context problem again and this time i asked ai about the ctx, cancel line
i got suggested the following minimal code change
dont do the defer cancel () with the factory, instead call cancel in case of error. this seems to work.
[https://github.com/devture/matrix-corporal/compare/main...Reis-A:matrix-corporal:fix/context-cancellation-RESTAPI](https://github.com/devture/matrix-corporal/compare/main...Reis-A:matrix-corporal:fix/context-cancellation-RESTAPI)

u can build your own docker image calling with the downloaded code base running make build-container-image in the matrix-corporal folder

**Comment by spantaleev** [comment]
Thanks for the hint as to where the issue is!

I think I've done a better fix in https://github.com/devture/matrix-corporal/commit/0c27562c2d1c621c251532054f6b358696d3495f
This fix is part of the new matrix-corporal [3.1.7](https://github.com/devture/matrix-corporal/blob/e9f2c38a3eb870e0b116422e63c047d5d5d41961/CHANGELOG.md#version-317-2025-09-20) release.
